pkg/provisioner/ironic/ironic.go

```diff
@@ -303,18 +304,20 @@ func (p *ironicProvisioner) findExistingHost() (ironicNode *nodes.Node, err erro
 	}
 
 	// Try to load the node by name
-	p.log.Info("looking for existing node by name", "name", p.host.Name)
-	ironicNode, err = nodes.Get(p.client, p.host.Name).Extract()
-	switch err.(type) {
-	case nil:
-		p.log.Info("found existing node by name")
-		return ironicNode, nil
-	case gophercloud.ErrDefault404:
-		p.log.Info(
-			fmt.Sprintf("node with name %s doesn't exist", p.host.Name))
-	default:
-		return nil, errors.Wrap(err,
-			fmt.Sprintf("failed to find node by name %s", p.host.Name))
+	for _, nodeName := range []string{p.ironicNodeNameFromHost(), p.host.Name} {
```

Comment:
Thinking a bit more about this, I think I was wrong and we don't reeeeally need the fallback in practice.

* If the Ironic DB is in ephemeral storage (almost everyone) then when you update all of the old nodes will be lost anyway. The only exception would be if you change the config but the Ironic DB is in a different Pod that doesn't get restarted.
* If it's in persistent storage, you won't run into this except for Hosts that were just created in the middle of switching from namespaced->clusterwide and which failed to write their ProvisionerID to the k8s API.

However, there is a real security issue here. If you have RBAC permissions to create a BareMetalHost, even in a namespace where it is not reconciled, then you could create a Host with the name `admin-namespace~admin-host` in advance of the switch from single-namespace to cluster-wide. Then you will end up with a BareMetalHost that has the same ProvisionerID as the one in the admin namespace, and thus with some luck and maybe timing you will be able to control it without having the credentials.

We could eliminate the security issue by not falling back in cases where the host name contains a `~`.

Reply:
done